akka-typed-tests/src/test/java/akka/typed/javadsl/BehaviorChainTest.java

```diff
@@ -0,0 +1,45 @@
+/**
+ * Copyright (C) 2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.typed.javadsl;
+
+import akka.typed.Behavior;
+import akka.typed.Terminated;
+import org.junit.Test;
+import org.scalatest.junit.JUnitSuite;
+
+import java.util.ArrayList;
+
+import static akka.typed.javadsl.Actor.same;
+import static akka.typed.javadsl.Actor.stopped;
+
+public class BehaviorChainTest extends JUnitSuite {
+    interface Message {
+    }
+
+    static final class One implements Message {
+        public String foo() {
+          return "Bar";
+        }
+    }
+    static final class MyList<T> extends ArrayList<T> implements Message {
+    };
+
+    @Test
+    public void shouldCompile() {
+      Behavior<Message> b = Actor.<Message>immutable()
```

Comment:
I think there can be advantages of having a real builder, such as performance optimizations. E.g. if the builder only contains exact message class matches it can can use a HashMap instead of traversing the matches.

Another thing we must consider is the order of the matches. The above example will match on `startsWith("a")` first even though we read it from top-down. I think we must use the same order as in the AbstractActor `PFBuilder`. With a real builder we have more freedom for efficiently implementing such things also.